### PR TITLE
Initial Implementation for GooglePubSub Connector

### PR DESCRIPTION
**Pull request by ayodhyamanewa**
## User stories
User Story of GooglePubSub : https://docs.google.com/document/d/1n6nOpaPQFHLWN4WrGEfUUCWivUQ1g9-Utf7IGJMBlwg/edit

## Automation tests
 - Unit tests 
   > Code coverage information = 82%
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

 operating systems = ubuntu 16.04 LTS

### CODE REVIEW

**Comment by grainier** [comment]
@ayodhyamanewa Please fill in the PR template with a few details on what this PR about, 
Purpose, Approach, User stories, Automation test, etc...

File: component/src/main/java/org/wso2/extension/siddhi/io/googlepubsub/sink/GooglePubSubSink.java [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+            FileInputStream serviceAccountStream = new FileInputStream(credentialsPath);
+            credentials = ServiceAccountCredentials.fromStream(serviceAccountStream);
+        } catch (IOException e) {
+            throw new SiddhiAppCreationException("The file that contains your service account credentials is not "
+                    + "found or you are not permitted to make authenticated calls. Check the credential.path '"
+                    + credentialPath + "' defined in stream " + siddhiAppName + " : " + streamID + ".", e);
+        }
+        createTopic();
+        try {
+            publisher = Publisher.newBuilder(topic).setCredentialsProvider(FixedCredentialsProvider.create(credentials))
```
Comment (tishan89) [inline component/src/main/java/org/wso2/extension/siddhi/io/googlepubsub/sink/GooglePubSubSink.java]: Does it make sense to retry this request assuming failure occurred due to network/connectivity issue. If so we need to move this block to connect()

File: component/src/main/java/org/wso2/extension/siddhi/io/googlepubsub/sink/GooglePubSubSink.java [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+
+    @Override
+    public void disconnect() {
+
+        if (publisher != null) {
+            try {
+                publisher.shutdown();
+                publisher.awaitTermination(1, TimeUnit.MINUTES);
+            } catch (Exception e) {
+                log.error("Error in shutting down the publisher : " + publisher);
```
Comment (tishan89) [inline component/src/main/java/org/wso2/extension/siddhi/io/googlepubsub/sink/GooglePubSubSink.java]: Print e.getMessage() also

File: component/src/main/java/org/wso2/extension/siddhi/io/googlepubsub/sink/GooglePubSubSink.java [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+     */
+    private void createTopic() {
+
+        try {
+            TopicAdminSettings topicAdminSettings = TopicAdminSettings.newBuilder().setCredentialsProvider
+                    (FixedCredentialsProvider.create(credentials)).build();
+            topicAdminClient = TopicAdminClient.create(topicAdminSettings);
+            topicAdminClient.createTopic(topic);
+        } catch (ApiException e) {
+            if (e.getStatusCode().getCode() != StatusCode.Code.ALREADY_EXISTS) {
```
Comment (tishan89) [inline component/src/main/java/org/wso2/extension/siddhi/io/googlepubsub/sink/GooglePubSubSink.java]: Check for topic existence and log a message if that is the case

File: component/src/main/java/org/wso2/extension/siddhi/io/googlepubsub/source/GooglePubSubMessageReceiver.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+                condition.await();
+            } catch (InterruptedException ie) {
+                Thread.currentThread().interrupt();
+            } finally {
+                lock.unlock();
+            }
+        }
+
+        ackReplyConsumer.ack();
+        sourceEventListener.onEvent("message :" + "\"" + pubsubMessage.getData().toStringUtf8() + "\"" + " ",
```
Comment (tishan89) [inline component/src/main/java/org/wso2/extension/siddhi/io/googlepubsub/source/GooglePubSubMessageReceiver.java]: Why does we need slashes

File: component/src/main/java/org/wso2/extension/siddhi/io/googlepubsub/source/GooglePubSubMessageReceiver.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+                null);
+    }
+
+    void pause() {
+
```
Comment (tishan89) [inline component/src/main/java/org/wso2/extension/siddhi/io/googlepubsub/source/GooglePubSubMessageReceiver.java:66]: Please remove unwanted space

File: component/src/main/java/org/wso2/extension/siddhi/io/googlepubsub/source/GooglePubSubMessageReceiver.java [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+        paused = true;
+    }
+
+    void resume() {
+
```
Comment (tishan89) [inline component/src/main/java/org/wso2/extension/siddhi/io/googlepubsub/source/GooglePubSubMessageReceiver.java:71]: Please remove unwanted space

File: component/src/main/java/org/wso2/extension/siddhi/io/googlepubsub/source/GooglePubSubSource.java [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+     */
+    private void createSubscription() {
+
+        try {
+            SubscriptionAdminSettings subscriptionAdminSettings = SubscriptionAdminSettings.newBuilder()
+                    .setCredentialsProvider(FixedCredentialsProvider.create(credentials)).build();
+            subscriptionAdminClient = SubscriptionAdminClient.create(subscriptionAdminSettings);
+            subscriptionAdminClient.createSubscription(subscriptionName, topicName, PushConfig.getDefaultInstance(),
+                    10);
+        } catch (ApiException e) {
```
Comment (tishan89) [inline component/src/main/java/org/wso2/extension/siddhi/io/googlepubsub/source/GooglePubSubSource.java]: Log a message if subscription already exists

File: component/src/main/resources/META-INF/services/io.grpc.LoadBalancerProvider [hunk 8]
```diff
@@ -0,0 +1,2 @@
+io.grpc.internal.PickFirstLoadBalancerProvider
```
Comment (tishan89) [inline component/src/main/resources/META-INF/services/io.grpc.LoadBalancerProvider:1]: remove above binary files

File: component/src/main/resources/log4j.properties [hunk 9]
```diff
@@ -0,0 +1,17 @@
+# For the general syntax of property based configuration files see the
```
Comment (tishan89) [inline component/src/main/resources/log4j.properties:18]: add licence here

File: component/src/test/java/org/wso2/extension/siddhi/io/googlepubsub/sink/TestCaseOfGooglePubSubSink.java [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+
+public class TestCaseOfGooglePubSubSink {
+
+    private static Logger log = Logger.getLogger(TestCaseOfGooglePubSubSink.class);
+    private volatile int count;
+    private volatile boolean eventArrived;
+
+    @BeforeMethod
+    public void initBeforeMethod() {
+
```
Comment (tishan89) [inline component/src/test/java/org/wso2/extension/siddhi/io/googlepubsub/sink/TestCaseOfGooglePubSubSink.java]: Please remove additional space

File: component/src/test/resources/log4j.properties [hunk 11]
```diff
@@ -0,0 +1,14 @@
+# For the general syntax of property based configuration files see the
```
Comment (tishan89) [inline component/src/test/resources/log4j.properties:1]: add licence here

File: component/src/test/resources/security/sp.json [hunk 12]
```diff
@@ -0,0 +1,12 @@
+{
```
Comment (tishan89) [inline component/src/test/resources/security/sp.json:1]: remove this file

File: docs/api/1.0.0-SNAPSHOT.md [hunk 13]
```diff
@@ -0,0 +1,129 @@
+# API Docs - v1.0.0-SNAPSHOT
```
Comment (tishan89) [inline docs/api/1.0.0-SNAPSHOT.md]: Remove this file and above empty file

File: pom.xml [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+            </dependency>
+            <dependency>
+                <groupId>org.wso2.extension.siddhi.map.text</groupId>
+                <artifactId>siddhi-map-text</artifactId>
+                <version>${text.mapper.version}</version>
+            </dependency>
+            <dependency>
+                <groupId>org.testng</groupId>
+                <artifactId>testng</artifactId>
+                <version>6.11</version>
```
Comment (tishan89) [inline pom.xml]: Move version to property

**Review by tishan89 — CHANGES_REQUESTED** [review]
Apart from requested changes please also make changes to test running as discusssed

File: component/src/main/java/org/wso2/extension/siddhi/io/googlepubsub/source/GooglePubSubMessageReceiver.java (hunk 5 above)
Comment (ayodhyamanewa) [reply component/src/main/java/org/wso2/extension/siddhi/io/googlepubsub/source/GooglePubSubMessageReceiver.java:66]: when we run wso2 code-style that line added automatically.

File: component/src/main/java/org/wso2/extension/siddhi/io/googlepubsub/source/GooglePubSubMessageReceiver.java (hunk 6 above)
Comment (ayodhyamanewa) [reply component/src/main/java/org/wso2/extension/siddhi/io/googlepubsub/source/GooglePubSubMessageReceiver.java:71]: when we run wso2 code-style that line added automatically.

File: component/src/test/java/org/wso2/extension/siddhi/io/googlepubsub/sink/TestCaseOfGooglePubSubSink.java (hunk 10 above)
Comment (ayodhyamanewa) [reply component/src/test/java/org/wso2/extension/siddhi/io/googlepubsub/sink/TestCaseOfGooglePubSubSink.java]: when we run wso2 code-style that line added automatically.

File: component/src/main/resources/META-INF/services/io.grpc.LoadBalancerProvider (hunk 8 above)
Comment (ayodhyamanewa) [reply component/src/main/resources/META-INF/services/io.grpc.LoadBalancerProvider:1]: When I remove those binary files, I get an issue saying no functional service provider and try adding a dependency okhttp or netty

File: component/src/test/resources/security/sp.json (hunk 12 above)
Comment (ayodhyamanewa) [reply component/src/test/resources/security/sp.json:1]: This is a dummy file with wrong credentials.

**Review by tishan89 — APPROVED** [review]

**Pull request merged**